engine/src/state_chain_observer/sc_observer/mod.rs

```diff
@@ -529,3 +524,80 @@ where
         Err(anyhow!("State Chain block stream ended"))
     }.instrument(info_span!("SCObserver")).boxed()).await
 }
+
+pub struct CeremonyIdCounters {
+	pub ethereum: CeremonyId,
+	pub polkadot: CeremonyId,
+	pub bitcoin: CeremonyId,
+}
+
+/// Get the ceremony id counters for each chain at the **start** of the given block without
+/// accessing the previous block
+pub async fn get_ceremony_id_counters_before_block<StateChainClient>(
+	block_hash: state_chain_runtime::Hash,
+	state_chain_client: Arc<StateChainClient>,
+) -> Result<CeremonyIdCounters, anyhow::Error>
+where
+	StateChainClient: SignedExtrinsicApi
+		+ UnsignedExtrinsicApi
+		+ 'static
+		+ Send
+		+ Sync
+		+ super::client::storage_api::StorageApi,
+{
+	let events: Vec<_> = state_chain_client
+		.storage_value::<pallet_cf_cfe_interface::CfeEvents<Runtime>>(block_hash)
+		.await
+		.map_err(|error| anyhow!("Failed to decode events at block hash {block_hash}. {error}"))?;
+
+	// Find the first event for each chain that contains a ceremony id and subtract 1 from it. If
+	// none is found, use the ceremony id counter from storage (ceremony id did not change during
+	// this block).
+	Ok(CeremonyIdCounters {
+		ethereum: if let Some(ceremony_id) = events.iter().find_map(|event| match event {
+			CfeEvent::EthThresholdSignatureRequest(req) => Some(req.ceremony_id),
+			CfeEvent::EthKeygenRequest(req) => Some(req.ceremony_id),
+			_ => None,
+		}) {
+			ceremony_id.saturating_sub(1)
+		} else {
+			state_chain_client
+				.storage_value::<pallet_cf_vaults::CeremonyIdCounter<
+					state_chain_runtime::Runtime,
+					state_chain_runtime::EthereumInstance,
+				>>(block_hash)
+				.await
+				.context("Failed to get Ethereum CeremonyIdCounter from SC")?
+		},
+		polkadot: if let Some(ceremony_id) = events.iter().find_map(|event| match event {
+			CfeEvent::DotThresholdSignatureRequest(req) => Some(req.ceremony_id),
+			CfeEvent::DotKeygenRequest(req) => Some(req.ceremony_id),
+			_ => None,
+		}) {
+			ceremony_id.saturating_sub(1)
+		} else {
+			state_chain_client
+				.storage_value::<pallet_cf_vaults::CeremonyIdCounter<
+					state_chain_runtime::Runtime,
+					state_chain_runtime::PolkadotInstance,
+				>>(block_hash)
+				.await
+				.context("Failed to get Polkadot CeremonyIdCounter from SC")?
+		},
+		bitcoin: if let Some(ceremony_id) = events.iter().find_map(|event| match event {
+			CfeEvent::BtcThresholdSignatureRequest(req) => Some(req.ceremony_id),
+			CfeEvent::BtcKeygenRequest(req) => Some(req.ceremony_id),
```

Comment:
I think you need to include key handover request here too.

Reply:
Good catch. I changed the unit test to cover this sort of mistake.